pkg/controller/s3/bucket/notificationConfig_test.go

```diff
@@ -378,3 +379,235 @@ func TestNotifLateInit(t *testing.T) {
+						{
+							Events:            generateNotificationAWSEvents(),
+							Filter:            generateAWSNotificationFilter(),
+							Id:                awsclient.String("test-id-2"),
+							LambdaFunctionArn: awsclient.String("lambda:321"),
+						},
+						{
+							Events:            generateNotificationAWSEvents(),
+							Filter:            generateAWSNotificationFilter(),
+							Id:                &id,
+							LambdaFunctionArn: &lambdaArn,
+						}},
+				},
+			},
+			want: want{
+				isUpToDate: 0,
+			},
+		},
+		"IsUpToDateEmpty": {
+			args: args{
+				cr: &v1beta1.NotificationConfiguration{},
+				b: &s3.GetBucketNotificationConfigurationOutput{
+					LambdaFunctionConfigurations: []s3types.LambdaFunctionConfiguration{},
+					QueueConfigurations:          []s3types.QueueConfiguration{},
+					TopicConfigurations:          []s3types.TopicConfiguration{},
+				},
+			},
+			want: want{
+				isUpToDate: 0,
+			},
+		},
+		"IsUpToDateNilInput": {
+			args: args{
+				cr: nil,
+				b: &s3.GetBucketNotificationConfigurationOutput{
+					LambdaFunctionConfigurations: []s3types.LambdaFunctionConfiguration{},
+					QueueConfigurations:          []s3types.QueueConfiguration{},
+					TopicConfigurations:          []s3types.TopicConfiguration{},
+				},
+			},
+			want: want{
+				isUpToDate: 0,
+			},
+		},
+		"IsUpToDateNilInputNeedsDelete": {
+			args: args{
+				cr: nil,
+				b: &s3.GetBucketNotificationConfigurationOutput{
+					LambdaFunctionConfigurations: []s3types.LambdaFunctionConfiguration{{
+						Events:            generateNotificationAWSEvents(),
+						Filter:            generateAWSNotificationFilter(),
+						Id:                &id,
+						LambdaFunctionArn: &lambdaArn,
+					}},
+				},
+			},
+			want: want{
+				isUpToDate: 2,
+			},
+		},
+		"IsUpToDateFalseMissing": {
+			args: args{
+				cr: &v1beta1.NotificationConfiguration{
+					LambdaFunctionConfigurations: []v1beta1.LambdaFunctionConfiguration{{
+						Events:            generateNotificationEvents(),
+						Filter:            generateNotificationFilter(),
+						ID:                &id,
+						LambdaFunctionArn: lambdaArn,
+					}},
+				},
+				b: &s3.GetBucketNotificationConfigurationOutput{},
+			},
+			want: want{
+				isUpToDate: 1,
+			},
+		},
+		"IsUpToDateExtraNeedsDeletion": {
+			args: args{
+				cr: &v1beta1.NotificationConfiguration{},
+				b: &s3.GetBucketNotificationConfigurationOutput{
+					LambdaFunctionConfigurations: []s3types.LambdaFunctionConfiguration{{
+						Events:            generateNotificationAWSEvents(),
+						Filter:            generateAWSNotificationFilter(),
+						Id:                &id,
+						LambdaFunctionArn: &lambdaArn,
+					}},
+				},
+			},
+			want: want{
+				isUpToDate: 2,
+			},
+		},
+	}
+
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			g := NewGomegaWithT(t)
+			actual, err := IsNotificationConfigurationUpToDate(tc.args.cr, tc.args.b)
+			g.Expect(err).NotTo(HaveOccurred())
```

Comment:
We try to avoid Gomega usage in our tests. I want to include this with the v0.21.1 release today so I plan to merge and quickly follow up to remove this.